rfcs/wptserve_py3.md

```diff
@@ -0,0 +1,121 @@
+
+**Size of codebase (as of 2020-03-30 revision 4e3194253a):**
+
+*   /tools/wptserve: ~7000 lines
+*   /tools/serve: ~1000 lines
+*   Custom handlers (test cases): 400+ custom handlers, ~20,000 lines \
+This is a rough count of all Python files excluding some directories, which still includes some util scripts: `find . -name "*.py" \! -path "./tools/*" \! -path "./webdriver/*" \! -path "./docs/*" \! -path "./websockets/*" \! -path "./resources/*"`
+
+
+## Goals
+
+Roughly in the order of importance:
+
+1.  Everything, including wptserve and custom handlers, must continue to work in both Python 2 and 3. (This was discussed and agreed upon at TPAC 2019.)
+1.  Custom handlers must continue to be able to read and write raw bytes from requests and to responses, including headers and body. wptserve must not change the byte sequences in any way; no errors should be raised if the byte sequences do not have a valid encoding.
+1.  _Optional/competing_: (Which goal to prioritize largely determines which option we choose.)
+    a. Keep type semantics consistent across Python 2 and 3 (a string is either always binary or always text).
+    a. Minimize code change to custom handlers, and consequently minimize the burden on test authors in the future by imposing fewer new requirements.
+
+
+## Root problem (or; why is this hard?)
+
+In both Python 2 and 3, unprefixed string literals have the `str` type. However, the meaning of the `str` type has changed: in Python 2 it is a raw byte sequence, whereas in Python 3 it is a Unicode text string. The rationale behind the change was to move to a Unicode-by-default world: arguably most people who used `str` in Python 2 were probably dealing with text and should use Unicode, so this decision minimized the impact on code handling human-readable text. However in WPT we actually intended to use byte sequences (see [background](#background)).
+
+This pain is further compounded by the choice of types in some parts of the Python 3 standard library. The low-level [`http`](https://docs.python.org/3/library/http.html) library takes and returns the `str` type (i.e. Unicode string) almost everywhere except the body (see the [type](https://github.com/python/typeshed/blob/master/stdlib/3/http/client.pyi) [annotations](https://github.com/python/typeshed/blob/master/stdlib/3/http/server.pyi)). This is convenient for existing code from Python 2 but semantically different and makes it difficult to send raw bytes on the wire. On the upside, the standard library carefully and intentionally uses an 8-bit isomorphic encoding (latin-1/ISO-8859-1) when appropriate (e.g. [headers](https://github.com/python/cpython/blob/8c3ab189ae552401581ecf0b260a96d80dcdae28/Lib/http/client.py#L220)) to allow arbitrary byte sequences:
+
+*   Getting bytes out: “encode” the return value with latin-1
+*   Passing bytes in: “decode” the byte sequence with latin-1
+
+
+## Options
```

Comment:
For clarity, I think we want to split these more clearly. This one should be something like `Proposed solution` (top-level, replacing 'Options'), and the other can move down to an Appendix or Alternatives with a note that it was not chosen (and why).